### ISSUE TITLE
Adaptive Gradient Clipping

### ISSUE THREAD

**Issue by edenlightning** [PyTorchLightning/pytorch-lightning#2963]
## 🚀 Feature
See code here: https://github.com/pseeth/autoclip

### Motivation

a simple method for automatically and adaptively choosing a gradient clipping threshold, based on the history of gradient norms observed during training. Experimental results show that applying AutoClip results in improved generalization performance for audio source separation networks. Observation of the training dynamics of a separation network trained with and without AutoClip show that AutoClip guides optimization into smoother parts of the loss landscape. AutoClip is very simple to implement and can be integrated readily into a variety of applications across multiple domains.

**Comment by pseeth** [comment]
AutoClip author here, I'd be happy to contribute this to Lightning. I'll be able to carve out time to do this soon I think if there's interest. Let me know how to get started?

**Comment by clementpoiret** [comment]
Any news for AGC in Lightning?

**Comment by pseeth** [comment]
I wasn't sure what the API should look like exactly but would love to contribute it with some high-level guidance.

**Comment by clementpoiret** [comment]
I don't know if that's possible but maybe it could be added to the trainer like gradient clipping?
Gradient clipping is done through `Trainer(gradient_clip_val=0.5)`, so that we'd have something like `trainer = Trainer(adaptive_gradient_clipping=True)`

It could be an awesome addition I think

**Comment by edenlightning** [comment]
Hey There @pseeth! what are your thoughts on https://github.com/pytorchlightning/pytorch-lightning/issues/6346?

Would be great to collaborate with you on this!

**Comment by pseeth** [comment]
Oh this looks great! Great timing, I just started looking into this over the weekend. I tried to make the following modifications in this commit just to get a feel for the scope of it. I got this:

https://github.com/pseeth/pytorch-lightning/commit/f5a239c18ab752bee35e71a5a460c5980570f7ed

I think AutoClip would be a super useful high-level Trainer flag. It really just works for me, and prevents searching for gradient clipping parameters. I've had mine set to `percentile=10` for a year now in all my code and haven't had to tweak it at all. One requirement for AutoClip though is keeping the history of gradient norms in a list, so that you can compute the percentile of it. Looking into the codebase, I couldn't figure out how to keep track of that history, so any pointers there would be helpful. 

Other than that, I think the proposed changes in #6346 would be great. Then, all the history of the gradient norms model can just belong to that model. One issue with how it might interact with AutoClip is having multiple LightningModules operating together. Should each one keep its own gradient history and have its own AutoClip value? It's unclear to me. But that can perhaps be pushed to the user side of things, by making sure to group all their modules into a single LightningModule before passing it to the Trainer.

**Comment by p-sodmann** [comment]
@pseeth you did not include the compute_grad_norm function in your example. I threw in the version in your repository and got this as a working solution, maybe someone else also needs it before it is implemented correctly.

``` python
class AutoClip:
    def __init__(self, percentile):
        self.grad_history = []
        self.percentile = percentile
        
    def compute_grad_norm(self, model):
        total_norm = 0
        for p in model.parameters():
            if p.grad is not None:
                param_norm = p.grad.data.norm(2)
                total_norm += param_norm.item() ** 2
        total_norm = total_norm ** (1. / 2)
        
        return total_norm 

    def __call__(self, model):
        grad_norm = self.compute_grad_norm(model)
        self.grad_history.append(grad_norm)
        clip_value = np.percentile(self.grad_history, self.percentile)
        torch.nn.utils.clip_grad_norm_(model.parameters(), clip_value)  
```

then i added it to the lightningmodule and its training step

``` python
def training_step(self, batch, batch_idx):
        # training_step defines the train loop. It is independent of forward
        
        opt = self.optimizers()
        opt.zero_grad()
        
        x = batch["image"].float()
        y = batch["mask"].float()
        
        prediction = self.model(x)
        
        loss = self.criterion(prediction, y)
        
        self.manual_backward(loss)
        self.autoclip(model)

        opt.step()
```

**Comment by akanyaani** [comment]
You can try zclip https://github.com/bluorion-com/ZClip
